docs/proposals/service-external-name.md

```diff
@@ -0,0 +1,150 @@
+to some problem. Ranging from using another upstream DNS server to creating a
+Name object assoaciated with DNSs.
+
+# Proposed solution
+
+The proposed solution is to add this at the service layer by adding a new
+`externalName` type to the service. This will create a CNAME record in the
+internal cluster DNS service.
+
+Using a CNAME avoids having to do the lookup, decide a timeout for it, and
+having to lookup for it when the TTL expires. It's way simpler to implement,
+while solving the right problem. And doing it at the service layer avoids all
+the problem discussed with doing the change at the endpoints layer.
+
+The proposed solution is the one by Tim Hockin here:
+https://github.com/kubernetes/kubernetes/issues/13748#issuecomment-230397975
+
+Currently a ServiceSpec looks like this (comments stripped down):
+```
+type ServiceSpec struct {
+    Ports []ServicePort
+
+    // If not specified, the associated Endpoints object is not automatically managed
+    Selector map[string]string
+
+    // "", a real IP, or "None".  If not specified, this is default allocated.  If "None", this Service is not load-balanced
+    ClusterIP string
+
+    // ClusterIP, NodePort, LoadBalancer.  Only applies if clusterIP != "None"
+    Type ServiceType
+
+    // Only applies if clusterIP != "None"
+    ExternalIPs []string
+    SessionAffinity ServiceAffinity
+
+    // Only applies to type=LoadBalancer
+    LoadBalancerIP string
+    LoadBalancerSourceRanges []string
+```
+
+So, the proposal is to change it to:
+```
+type ServiceSpec struct {
+    Ports []ServicePort
+
+    // If not specified, the associated Endpoints object is not automatically managed
++   // Only applies if type is ClusterIP, NodePort, or LoadBalancer.  If type is ExternalName, this is ignored.
+    Selector map[string]string
+
+    // "", a real IP, or "None".  If not specified, this is default allocated.  If "None", this Service is not load-balanced.
++   // Only applies if type is ClusterIP, NodePort, or LoadBalancer.  If type is ExternalName, this is ignored.
+    ClusterIP string
+
+-   // ClusterIP, NodePort, LoadBalancer.  Only applies if clusterIP != "None"
++   // ExternalName, ClusterIP, NodePort, LoadBalancer.  Only applies if clusterIP != "None"
+    Type ServiceType
+
++   // Only applies if type is ExternalName
++   ExternalName string
+
+    // Only applies if clusterIP != "None"
+    ExternalIPs []string
+    SessionAffinity ServiceAffinity
+
+    // Only applies to type=LoadBalancer
+    LoadBalancerIP string
+    LoadBalancerSourceRanges []string
+```
+
+So, for example, you it can be used like this:
+```
+apiVersion: v1
+kind: Service
+metadata:
+  name: my-rds
+spec:
+  ports:
+  - port: 12345
+type: ExternalName
+externalName: myapp.rds.whatever.aws.says
+```
+
+There is one thing to take into account (that no other alternative considered
+fixes either): TLS.
+
+If the service is a CNAME for an endpoint that uses TLS, connecting with another
```

Comment:
On Sun, Jul 17, 2016 at 12:35:30PM -0700, Clayton Coleman wrote:
> > +```
> > +apiVersion: v1
> > +kind: Service
> > +metadata:
> > +  name: my-rds
> > +spec:
> > +  ports:
> > +  - port: 12345
> > +type: ExternalName
> > +externalName: myapp.rds.whatever.aws.says
> > +```
> > +
> > +There is one thing to take into account (that no other alternative considered
> > +fixes either): TLS.
> > +
> > +If the service is a CNAME for an endpoint that uses TLS, connecting with another
> 
> You can probably reframe this a bit - there are things we could do (like
> expose referential TLS reencrypt) but for which there is no immediate

What is to "expose referential TLS reencrypt"?

> situation.  I'd leave TLS as a "future consideration" and suggest a few of the
> options.

Cool, will change it. What are the options besides having the cert for the k8s
name service and failing TLS validation? So I can list them (not sure what
"expose referential TLS reencrypt" means :)